### PR TITLE
RNGManager access through Locator

### PR DESCRIPTION
**Pull request by FalseIlyu**
Simplest way to make the RNGManager access through Locator

### CODE REVIEW

File: src/Locator.h [hunk 1]
```diff
@@ -19,5 +20,6 @@ namespace openblack
 struct Locator
 {
 	using resources = entt::service_locator<resources::ResourcesInterface>;
+	using rng = entt::service_locator<RNGManager>;
```
Comment (raffclar) [inline src/Locator.h:23]: Create an interface for RNGManager called `RNGManagerInterface`. We don't want to couple the locator to an implementation.

File: src/Game.cpp [hunk 2]
```diff
@@ -418,6 +418,7 @@ bool Game::Update()
 bool Game::Initialize()
 {
 	Locator::resources::set<resources::Resources>();
+	Locator::rng::set<RNGManager>();
```
Comment (raffclar) [inline src/Game.cpp:421]: 👍

**Review by raffclar — CHANGES_REQUESTED** [review]

File: src/Common/RNG.h [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+		    std::conditional_t<std::is_integral_v<T>, std::uniform_int_distribution<T>, std::uniform_real_distribution<T>>;
+		dist_t dist(min, max);
+		if (isDebug())
+		{
+			std::lock_guard<std::mutex> safe_lock(lockAll());
+			return dist(generator());
+		}
+		return dist(generator());
+	}
+	virtual ~RNGInterface() {}
```
Comment (bwrsandman) [inline src/Common/RNG.h:34]: Move the implementation to the implementation file.
```
~RNGInterface::RNGInterface() = default;
```

File: src/Common/RNGManager.cpp [hunk 4]
```diff
@@ -16,16 +17,30 @@ RNGManager::RNGManager()
...
 	return _debugRng == isDebug;
 }
+
+std::mt19937& RNGManager::generator()
+{
+	if (isDebug())
+	{
+		return _generator;
+	}
+	thread_local std::mt19937 tgenerator(time(nullptr));
```
Comment (bwrsandman) [inline src/Common/RNGManager.cpp:34]: Fix these warning:
>  'argument': conversion from 'time_t' to 'unsigned int', possible loss of data

**Review by bwrsandman — CHANGES_REQUESTED** [review]
I feel like the design here is wrong.

If we're going through the Locator pattern, debug and non-debug should be two different implementations.
of `RNGInterface`.

**Review by bwrsandman — CHANGES_REQUESTED** [review]
Same as previously. With the locator pattern, it's best to have two implementations of RNG.h, which the name should be improved like: RandomNumberManager.h
It should provide two interfaces to the locator:
* RandomNumberManagerProduction.h
* RandomNumberManagerTesting.h

File: src/Game.cpp [hunk 5]
```diff
@@ -423,6 +424,7 @@ bool Game::Update()
...
```
Comment (bwrsandman) [inline src/Game.cpp:427]: This is where you would either set the debug or production rng implementation

File: src/Common/RandomNumberManagerProduction.cpp [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+std::mt19937& RandomNumberManagerProduction::generator()
+{
+	thread_local std::mt19937 tGenerator(static_cast<unsigned int>(time(nullptr)));
+	return tGenerator;
+}
+
+std::mutex& RandomNumberManagerProduction::lockAll()
+{
+	std::mutex rMutex;
+	return rMutex;
```
[the remark this answers is not in the document]
Comment (bwrsandman) [reply src/Common/RandomNumberManagerProduction.cpp:30]: Is a mutex necessary in production?

File: src/Common/RandomNumberManagerProduction.cpp [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+#include "RandomNumberManagerProduction.h"
+#include <ctime>
+
+using namespace openblack;
+
+bool RandomNumberManagerProduction::SetSeed(int seed)
+{
+	generator().seed(seed);
+	return true;
+}
```
Comment (bwrsandman) [inline src/Common/RandomNumberManagerProduction.cpp:19]: Do we need to manually set seed in production?

**Review by bwrsandman — CHANGES_REQUESTED** [review]

File: src/Common/RandomNumberManagerProduction.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (FalseIlyu) [reply src/Common/RandomNumberManagerProduction.cpp:30]: We need to have the lock callable for NextValue even if it is not used, it shouldn't be necessary for production but i don't quite know what could be done to have NextValue() be separate for test and production as most of the idea i found need for NextValue to not use template.

File: src/Common/RandomNumberManagerProduction.cpp (hunk 7 above)
Comment (bwrsandman) [reply src/Common/RandomNumberManagerProduction.cpp:19]: In that case wouldn't that be another implementation?
Maybe add a MultiplayerSynced implementation

File: src/Common/RandomNumberManager.h [hunk 8]
```diff
@@ -1,45 +1,43 @@
...
  * For a complete list of all authors, please refer to contributors.md
  * Interested in contributing? Visit https://github.com/openblack/openblack
  *
  * openblack is licensed under the GNU General Public License version 3.
- *****************************************************************************/
+ ******************************************************************************/
 
 #pragma once
 
+#include <functional>
```
Comment (bwrsandman) [inline src/Common/RandomNumberManager.h:12]: Are you sure you need this include?

**Review by bwrsandman — APPROVED** [review]
My issue remains that it would be ideal if we could have the RNG manager interface not require optional or mutex.

**Review by bwrsandman — CHANGES_REQUESTED** [review]
Please fix the warnings in `Files changed`

**Review by raffclar — APPROVED** [review]

**Review by bwrsandman — APPROVED** [review]

**Pull request merged**